### PR TITLE
enh: support type conversion from Integer to logical

### PR DESCRIPTION
**Pull request by HarshitaKalani**
Towards: #4017 
Fixes: #4041
Also a little cleanup of the code.

### CODE REVIEW

**Comment by HarshitaKalani** [comment]
I think we should throw a warning for this and rest of the conversions. Go ahead @kmr-srbh, please feel free to open a PR for the same.

**Comment by kmr-srbh** [comment]
> I think we should throw a warning for this and rest of the conversions. Go ahead @kmr-srbh, please feel free to open a PR for the same.

Sure! https://github.com/lfortran/lfortran/issues/3903 is related I believe?

**Comment by HarshitaKalani** [comment]
>https://github.com/lfortran/lfortran/issues/3903 is related I believe?
Yes.

**Comment by certik** [comment]
Is this needed for some code? I didn't know this is even allowed. It seems GFortran gives a warning. We should also give a warning at least, or possibly disable this unless a compiler option is given.

How does the ASR look like for the code above? Are the numbers `1` and `0` represented as integers, and then explicitly cast to logical? If you could post `--show-asr` here, that would be great.

In general, we need to be very careful with allowing such broad type conversions, because once we implement them, we can't take it back easily without breaking things.

If this is not in the standard, then I would not allow it by default: it should give an error message. We can then add a compiler option to enable this for codes that require it. Which codes require it? If we don't know codes that require it, then I would not even support this at all.

**Comment by HarshitaKalani** [comment]
```console
$ lfortran a.f90 --show-asr
(TranslationUnit
    (SymbolTable
        1
        {
            types_24:
                (Program
                    (SymbolTable
                        2
                        {
                            x:
                                (Variable
                                    2
                                    x
                                    []
                                    Local
                                    (Cast
                                        (ArrayConstant
                                            12
                                            [1, 1, 0]
                                            (Array
                                                (Integer 4)
                                                [((IntegerConstant 1 (Integer 4))
                                                (IntegerConstant 3 (Integer 4)))]
                                                FixedSizeArray
                                            )
                                            ColMajor
                                        )
                                        IntegerToLogical
                                        (Array
                                            (Logical 4)
                                            [((IntegerConstant 1 (Integer 4))
                                            (IntegerConstant 3 (Integer 4)))]
                                            FixedSizeArray
                                        )
                                        (ArrayConstant
                                            12
                                            [.true., .true., .false.]
                                            (Array
                                                (Logical 4)
                                                [((IntegerConstant 1 (Integer 4))
                                                (IntegerConstant 3 (Integer 4)))]
                                                FixedSizeArray
                                            )
                                            ColMajor
                                        )
                                    )
                                    (ArrayConstant
                                        12
                                        [.true., .true., .false.]
                                        (Array
                                            (Logical 4)
                                            [((IntegerConstant 1 (Integer 4))
                                            (IntegerConstant 3 (Integer 4)))]
                                            FixedSizeArray
                                        )
                                        ColMajor
                                    )
                                    Save
                                    (Array
                                        (Logical 4)
                                        [((IntegerConstant 1 (Integer 4))
                                        (IntegerConstant 3 (Integer 4)))]
                                        FixedSizeArray
                                    )
                                    ()
                                    Source
                                    Public
                                    Required
                                    .false.
                                )
                        })
                    types_24
                    []
                    [(Print
                        [(Var 2 x)]
                        ()
                        ()
                    )]
                )
        })
    []
)
```

**Comment by HarshitaKalani** [comment]
>Are the numbers 1 and 0 represented as integers, and then explicitly cast to logical? 

Yes
I came across this while debugging an issue, however I realised later that this was not related to the issue I was solving.

**Comment by certik** [comment]
Let's add a new compiler flag `--logical-casting` and only allow this when it is on. Give an error message without the flag (and suggest to use the flag if this was intended).

**Comment by HarshitaKalani** [comment]
This is ready for a review

File: src/bin/lfortran.cpp [hunk 1]
```diff
@@ -2162,13 +2163,15 @@ int main_app(int argc, char *argv[]) {
         compiler_options.implicit_argument_casting = true;
         compiler_options.implicit_interface = true;
         compiler_options.print_leading_space = true;
+        compiler_options.logical_casting = true;
```
Comment (certik) [inline src/bin/lfortran.cpp:2166]: I think the standard does not allow it, correct? So this should be false.

File: src/bin/lfortran.cpp [hunk 2]
```diff
@@ -2162,13 +2163,15 @@ int main_app(int argc, char *argv[]) {
...
     } else if (arg_standard == "legacy") {
         // f23
         compiler_options.disable_style = true;
         compiler_options.implicit_typing = true;
         compiler_options.implicit_argument_casting = true;
         compiler_options.implicit_interface = true;
         compiler_options.print_leading_space = true;
+        compiler_options.logical_casting = true;
```
Comment (certik) [inline src/bin/lfortran.cpp:2174]: This I don't know yet, but I would set it false as well.

File: run_tests.py [hunk 3]
```diff
@@ -160,6 +161,16 @@ def is_included(backend):
             update_reference,
             verify_hash,
             extra_args)
+    
+    if ast_logical_casting:
+        run_test(
+            filename,
+            "ast_logical_casting",
+            "lfortran --show-ast --logical-casting --no-color {infile} -o {outfile}",
```
Comment (certik) [inline run_tests.py:169]: It seems `--show-ast` should always work. Shouldn't this be `--show-asr`?

**Review by certik — COMMENTED** [review]
I think this looks in the right direction. I left some comments.

File: run_tests.py (hunk 3 above)
Comment (HarshitaKalani) [reply run_tests.py:169]: Oh right. I'll update that. Thank you.

File: src/lfortran/semantics/ast_common_visitor.h [hunk 4]
```diff
@@ -2860,183 +2860,119 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
...
-                                                    is_int = false;
-                                                    break;
-                                                }
-                                            }
-                                            if (is_int) {
-                                                ASR::expr_t* array_const = ASRUtils::EXPR(ASRUtils::make_ArrayConstructor_t_util(al, a->base.base.loc, body.p, body.size(), real_type, a->m_storage_format));
-                                                cast->m_value = ASRUtils::expr_value(array_const);
-                                                value = cast->m_value;
+                                                    is_convertible = true;
+                                                } else break;
```
Comment (Shaikh-Ubaid) [inline src/lfortran/semantics/ast_common_visitor.h:2878]: This looks strange. Let's use the regular `else` with `{ }` as used at other places.

```suggestion
                                                } else { 
                                                    break;
                                                }
```

Similarly for other `else break;` in this PR.

File: tests/reference/ast_logical_casting-type_casting_01-93aa212.stdout [hunk 5]
```diff
@@ -0,0 +1,44 @@
+(TranslationUnit
```
Comment (Shaikh-Ubaid) [inline tests/reference/ast_logical_casting-type_casting_01-93aa212.stdout:1]: Is this reference output file (and the associated json file) extra/unexpected?

File: tests/tests.toml [hunk 6]
```diff
@@ -3935,3 +3935,12 @@ asr = true
...
 asr = true
+
+[[test]]
+filename = "logical_casting.f90"
+asr = true
+
+[[test]]
+filename = "errors/type_casting_01.f90"
+asr = true
+ast_logical_casting = true
```
Comment (Shaikh-Ubaid) [inline tests/tests.toml:3946]: If we are trying to generate an error for this, then I think we don't need `ast_logical_casting` for this.

```suggestion
```

File: tests/tests.toml [hunk 7]
```diff
@@ -3935,3 +3935,12 @@ asr = true
 [[test]]
 filename = "character_parameter_padding_trimming.f90"
 asr = true
+
+[[test]]
+filename = "logical_casting.f90"
+asr = true
```
Comment (Shaikh-Ubaid) [inline tests/tests.toml:3941]: Assuming this is a test for verifying working of `--logical-casting`, I think we should remove this from here and directly add it as an integration tests and provide the `--logical-casting` in `EXTRA_ARGS`.

File: tests/errors/type_casting_01.f90 [hunk 8]
```diff
@@ -0,0 +1,5 @@
+program type_casting_01
+    implicit none
+    logical :: x(3) = [1, 1, 0]
+    print*, x
```
Comment (Shaikh-Ubaid) [inline tests/errors/type_casting_01.f90:4]:
```suggestion
    print *, x
```

**Review by Shaikh-Ubaid — COMMENTED** [review]
I shared some comments. Good work in eliminating duplicate code in `ast_common_visitor` and making it more general!

**Comment by Shaikh-Ubaid** [comment]
Two minor concerns:
- There should just be one commit about [add and register test (https://github.com/lfortran/lfortran/pull/4042/commits/61e283b10b82e4db527d2819b0a02f3c3755fdf9), so last commit should be squashed to it
- [add asr_logical_castinf check for tests](https://github.com/lfortran/lfortran/pull/4042/commits/af6630be58e2467804b70420b61083c048a1b923), here it seems error in the commit message.


Either we clean the commit history (where we fix the above two) or we squash merge the PR. Let me know how you would like to proceed @HarshitaKalani.

**Review by Shaikh-Ubaid — APPROVED** [review]
Just the commit history needs to be cleaned or the PR squash merged. Rest is good. Nice work @HarshitaKalani ! Thanks for this!

File: src/bin/lfortran.cpp (hunk 1 above)
Comment (Shaikh-Ubaid) [reply src/bin/lfortran.cpp]:
```console
% cat examples/expr2.f90     
program expr2
implicit none

logical :: x(3) = [1, 1, 0]
print *, x

end program
% gfortran examples/expr2.f90
examples/expr2.f90:4:17:

    4 | logical :: x(3) = [1, 1, 0]
      |                 1
Warning: Extension: Conversion from INTEGER(4) to LOGICAL(4) at (1)
% ./a.out                    
 T T F
```

Seems like `gfortran` by default allows it, but emits a warning. I think it is fine to be `false` by default. We can always later change the initialization value (to whichever we find to be the most appropriate).

**Comment by HarshitaKalani** [comment]
>Let me know how you would like to proceed @HarshitaKalani.

Oh I see. I'll clean the commit history. I think that would be better. Thanks for bringing this to notice.

**Pull request merged**